### PR TITLE
Don't arm the handshake timer if there's no data

### PR DESCRIPTION
**Pull request by ianswett**
Rewrites a substantial portion of the Crypto Retransmission Timeout section and specifies that the handshake timer is not armed if the time threshold loss detection timer is set.

Fixes the portion of #1964 not fixed in #2281

### CODE REVIEW

File: draft-ietf-quic-recovery.md [hunk 1]
```diff
@@ -458,11 +458,11 @@ sent, then no alarm should be armed until data has been received from the
 client.
 
 Because the server could be blocked until more packets are received, the client
-MUST start the crypto retransmission timer even if there is no unacknowledged
-CRYPTO data.  If the timer expires and the client has no CRYPTO data to
-retransmit and does not have Handshake keys, it SHOULD send an Initial packet in
-a UDP datagram of at least 1200 bytes.  If the client has Handshake keys, it
-SHOULD send a Handshake packet.
+MUST start the crypto retransmission timer until it has 1RTT keys, even if there
```
Comment (marten-seemann) [inline draft-ietf-quic-recovery.md]: How would a lost CFIN be retransmitted?

File: draft-ietf-quic-recovery.md [hunk 2]
```diff
@@ -1151,7 +1152,8 @@ OnLossDetectionTimeout():
     DetectLostPackets(pn_space)
   // Retransmit crypto data if no packets were lost
   // and there are still crypto packets in flight.
-  else if (crypto packets are in flight):
+  else if (crypto packets are in flight &&
+           crypto data to send):
     // Crypto retransmission timeout.
     RetransmitUnackedCryptoData()
     crypto_count++
```
Comment (martinduke) [inline draft-ietf-quic-recovery.md]: else if (client && (no 1-RTT keys)
    send a 1200B packet
or something to that effect

File: draft-ietf-quic-recovery.md [hunk 3]
```diff
@@ -1114,7 +1114,8 @@ SetLossDetectionTimer():
     loss_detection_timer.update(loss_time)
     return
 
-  if (crypto packets are in flight):
+  if (crypto packets are in flight &&
+      crypto data to send):
```
Comment (martinduke) [inline draft-ietf-quic-recovery.md]: || (client && (no 1-RTT keys yet))

File: draft-ietf-quic-recovery.md (hunk 1 above)
Comment (martinduke) [inline draft-ietf-quic-recovery.md]: MUST start the crypto retransmission timer upon timeout....

In line 448: ack-eliciting crypto packet?

LIne 451: "if possible, and restart the crypto retransmission timer if there is unacknowledged data"

File: draft-ietf-quic-recovery.md [hunk 4]
```diff
@@ -458,11 +458,11 @@ sent, then no alarm should be armed until data has been received from the
...
 
 Because the server could be blocked until more packets are received, the client
-MUST start the crypto retransmission timer even if there is no unacknowledged
-CRYPTO data.  If the timer expires and the client has no CRYPTO data to
-retransmit and does not have Handshake keys, it SHOULD send an Initial packet in
-a UDP datagram of at least 1200 bytes.  If the client has Handshake keys, it
-SHOULD send a Handshake packet.
+MUST start the crypto retransmission timer until it has 1RTT keys, even if there
+is no unacknowledged CRYPTO data.  If the timer expires and the client has no
+CRYPTO data to retransmit and does not have Handshake keys, it SHOULD send an
```
Comment (martinduke) [inline draft-ietf-quic-recovery.md]: SHOULD -> MUST? If we don't do this, the connection could hang in a fairly common case.

File: draft-ietf-quic-recovery.md (hunk 1 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: It's not just timeout(ie: it needs to be set after an ACK is received), but I'll try to clarify.

line 448: The definition of crypto packet is that it has a CRYPTO frame in it, so it's always ack-eliciting.

Retransmitting the crypto data would naturally update the timer(see previous sentence) and there would be unacknowledged crypto data in that case, so I'm not sure your extra clause is clarifying or potentially confusing?

Comment (martinduke) [reply draft-ietf-quic-recovery.md]: There would be outstanding CRYPTO data, so it would restart the timer and then send.

File: draft-ietf-quic-recovery.md (hunk 4 above)
Comment (martinduke) [reply draft-ietf-quic-recovery.md]: ah you beat me to it.

File: draft-ietf-quic-recovery.md (hunk 3 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: That's in the the other PR: #2281

File: draft-ietf-quic-recovery.md (hunk 2 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: Also in the other PR.

File: draft-ietf-quic-recovery.md (hunk 1 above)
Comment (marten-seemann) [reply draft-ietf-quic-recovery.md]: You're right, but it's kind of hard to understand.
What about the following structure: "MUST start the crypto retransmission timer if there is outstanding CRYPTO data. It MUST also start the crypto retransmission timer until it has 1-RTT keys,..."

File: draft-ietf-quic-recovery.md [hunk 5]
```diff
@@ -435,7 +435,9 @@ Data in CRYPTO frames is critical to QUIC transport and crypto negotiation, so a
 more aggressive timeout is used to retransmit it.
 
 The initial crypto retransmission timeout SHOULD be set to twice the initial
-RTT.
+RTT.  On each consecutive expiration of the crypto timer without receiving an
+acknowledgement for a new packet, the sender SHOULD double the crypto
+retransmission timeout and set a timer for this period.
```
Comment (martinduke) [inline draft-ietf-quic-recovery.md]: Not quite right. If we got some acknowledgments, we still need to restart the timer. And note the pseudocode comments below.

Comment (ianswett) [reply draft-ietf-quic-recovery.md]: This was just a reorg.  This text is about doubling the timer after expiry when no ACKs are received.

I think the pseudocode comments are all addressed in the other PR.

File: draft-ietf-quic-recovery.md (hunk 1 above)
Comment (martinduke) [reply draft-ietf-quic-recovery.md]: You're right, please disregard.

File: draft-ietf-quic-recovery.md [hunk 6]
```diff
@@ -458,17 +460,15 @@ sent, then no alarm should be armed until data has been received from the
...
-CRYPTO data.  If the timer expires and the client has no CRYPTO data to
-retransmit and does not have Handshake keys, it SHOULD send an Initial packet in
-a UDP datagram of at least 1200 bytes.  If the client has Handshake keys, it
-SHOULD send a Handshake packet.
-
-On each consecutive expiration of the crypto timer without receiving an
-acknowledgement for a new packet, the sender SHOULD double the crypto
-retransmission timeout and set a timer for this period.
+MUST ensure the crypto retransmission timer is set if there is unacknowledged
+crypto data and MUST ensure  the timer is set until it has 1RTT keys.
```
Comment (martinduke) [inline draft-ietf-quic-recovery.md]: double space

File: draft-ietf-quic-recovery.md (hunk 5 above)
Comment (martinduke) [reply draft-ietf-quic-recovery.md]: Ah, OK. Line 448 probably should read "When an acknowledgment is received, a new RTT is computed. The endpoint MUST restart the timer and SHOULD set it for twice the newly computed smoothed RTT."

File: draft-ietf-quic-recovery.md [hunk 7]
```diff
@@ -458,17 +462,16 @@ sent, then no alarm should be armed until data has been received from the
...
-SHOULD send a Handshake packet.
-
-On each consecutive expiration of the crypto timer without receiving an
-acknowledgement for a new packet, the sender SHOULD double the crypto
-retransmission timeout and set a timer for this period.
+MUST ensure the crypto retransmission timer is set if there is unacknowledged
+crypto data and MUST ensure the timer is set until it has 1-RTT keys.
+If the timer expires and the client has no CRYPTO data to retransmit and does
+not have Handshake keys, it MUST send an Initial packet in a UDP datagram of
+at least 1200 bytes.  If the client has Handshake keys, it MUST send a
```
Comment (marten-seemann) [inline draft-ietf-quic-recovery.md]: Why would the client send a Handshake packet if there’s no CRYPTO data to retransmit? If there’s no Handshake CRYPTO data to retransmit, this means that the server already received the client‘s TLS messages sent with Handshake keys, and won’t be blocked by the 3x Limit any more, doesn’t it?

**Comment by martinduke** [comment]
If the server first flight does not include finished and the client
handshake ack is lost, the client has no unacked data but the server is
flow control limited.

File: draft-ietf-quic-recovery.md (hunk 7 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: I believe the particular concern is that the client could have Handshake keys because they receive the server's Initial, but not receive the server's Handshake packet, so have nothing to send in Handshake encryption.

If the client ever had CRYPTO data to send in Handshake, then this would not be a concern, as you state.

File: draft-ietf-quic-recovery.md [hunk 8]
```diff
@@ -442,13 +442,17 @@ connections over the same network SHOULD use the previous connection's final
 smoothed RTT value as the resumed connection's initial RTT.  If no previous RTT
 is available, or if the network changes, the initial RTT SHOULD be set to 500ms,
 resulting in a 1 second initial handshake timeout as recommended in
-{{?RFC6298}}. When an acknowledgement is received, a new RTT is computed and the
-timer SHOULD be set for twice the newly computed smoothed RTT.
+{{?RFC6298}}. When the first acknowledgement is received, an RTT is computed and
+the timer SHOULD be set for twice the newly computed RTT.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: How about "When the first acknowledgement is received, the timer SHOULD be set for twice the value of the resulting RTT sample"? Now that I see further below, why is this necessary at all? I would remove this last sentence. The beginning of the next para says exactly this, since the smoothed_rtt is set to the RTT sample on the first ack.

File: draft-ietf-quic-recovery.md [hunk 9]
```diff
@@ -442,13 +442,17 @@ connections over the same network SHOULD use the previous connection's final
...
 
-When a crypto packet is sent, the sender MUST set a timer for the crypto
-timeout period.  This timer MUST be updated when a new crypto packet is sent.
-Upon timeout, the sender MUST retransmit all unacknowledged CRYPTO data if
-possible.
+When a crypto packet is sent, the sender MUST set a timer for twice the smoothed
+RTT.  This timer MUST be updated when a new crypto packet is sent and when
+an acknowledgement is received which computes a new RTT sample. Upon timeout,
+the sender MUST retransmit all unacknowledged CRYPTO data if possible.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: Why "if possible"?

File: draft-ietf-quic-recovery.md [hunk 10]
```diff
@@ -442,13 +442,17 @@ connections over the same network SHOULD use the previous connection's final
...
+
+On each consecutive expiration of the crypto timer without receiving an
+acknowledgement for a new packet, the sender SHOULD double the crypto
+retransmission timeout and set a timer for this period.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: I'd make this a MUST.

File: draft-ietf-quic-recovery.md [hunk 11]
```diff
@@ -458,17 +462,16 @@ sent, then no alarm should be armed until data has been received from the
...
-CRYPTO data.  If the timer expires and the client has no CRYPTO data to
-retransmit and does not have Handshake keys, it SHOULD send an Initial packet in
-a UDP datagram of at least 1200 bytes.  If the client has Handshake keys, it
-SHOULD send a Handshake packet.
-
-On each consecutive expiration of the crypto timer without receiving an
-acknowledgement for a new packet, the sender SHOULD double the crypto
-retransmission timeout and set a timer for this period.
+MUST ensure the crypto retransmission timer is set if there is unacknowledged
+crypto data and MUST ensure the timer is set until it has 1-RTT keys.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: This needs rephrasing. How about "MUST ensure that the crypto retransmission timer is set if there is unacknowledged crypto data or if the client does not yet have 1-RTT keys."

File: draft-ietf-quic-recovery.md [hunk 12]
```diff
@@ -458,17 +462,16 @@ sent, then no alarm should be armed until data has been received from the
...
+If the timer expires and the client has no CRYPTO data to retransmit and does
+not have Handshake keys, it MUST send an Initial packet in a UDP datagram of
+at least 1200 bytes.  If the client has Handshake keys, it MUST send a
+Handshake packet.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: Suggested reword: "If the crypto retransmission timer expires before the client has 1-RTT keys, it is possible that the client may not have any crypto data to retransmit. The client MUST still send a new packet to allow the server to continue sending data (see Section XX in {{QUIC-TRANSPORT}}).  If Handshake keys are available to the client, it MUST send a Handshake packet, and otherwise, an Initial packet in a UDP datagram of at least 1200 bytes."

File: draft-ietf-quic-recovery.md [hunk 13]
```diff
@@ -1114,7 +1117,8 @@ SetLossDetectionTimer():
     loss_detection_timer.update(loss_time)
     return
 
-  if (crypto packets are in flight):
+  if (crypto packets are in flight &&
+      crypto data to send):
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: what does this predicate mean? That there's new data to send?

File: draft-ietf-quic-recovery.md (hunk 8 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: SG, removed.

File: draft-ietf-quic-recovery.md (hunk 9 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: It can't if the anti-amplification limit limits it.

File: draft-ietf-quic-recovery.md (hunk 13 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: It's supposed to mean that there's unacknowledged crypto data.  Should I say "unacknowledged crypto data" instead?

File: draft-ietf-quic-recovery.md (hunk 12 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: Done, with some minor editing.

File: draft-ietf-quic-recovery.md (hunk 13 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: I removed this from here due to the other PR, but left it in below.  Feel free to suggest alternate phrasing.

File: draft-ietf-quic-recovery.md [hunk 14]
```diff
@@ -458,17 +463,18 @@ sent, then no alarm should be armed until data has been received from the
...
-SHOULD send a Handshake packet.
-
-On each consecutive expiration of the crypto timer without receiving an
-acknowledgement for a new packet, the sender SHOULD double the crypto
-retransmission timeout and set a timer for this period.
+MUST ensure that the crypto retransmission timer is set if there is
+unacknowledged crypto data or if the client does not yet have 1-RTT keys.
+If the crypto retransmission timer expires before the client has 1-RTT keys,
+it is possible that the client may not have any crypto data to retransmit.
+The client MUST send a new packet to allow the server to continue sending
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
However, the client MUST send a new packet, containing only PING or PADDDING frames if necessary, to allow the server to continue sending
```

File: draft-ietf-quic-recovery.md [hunk 15]
```diff
@@ -442,13 +442,18 @@ connections over the same network SHOULD use the previous connection's final
...
 
-When a crypto packet is sent, the sender MUST set a timer for the crypto
-timeout period.  This timer MUST be updated when a new crypto packet is sent.
-Upon timeout, the sender MUST retransmit all unacknowledged CRYPTO data if
-possible.
+When a crypto packet is sent, the sender MUST set a timer for twice the smoothed
+RTT.  This timer MUST be updated when a new crypto packet is sent and when
+an acknowledgement is received which computes a new RTT sample. Upon timeout,
+the sender MUST retransmit all unacknowledged CRYPTO data if possible.  The
+sender MUST NOT declare in flight crypto packets as lost when the crypto timer
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]:
```suggestion
sender MUST NOT declare in-flight crypto packets as lost when the crypto timer
```

File: draft-ietf-quic-recovery.md [hunk 16]
```diff
@@ -1151,7 +1158,8 @@ OnLossDetectionTimeout():
     DetectLostPackets(pn_space)
   // Retransmit crypto data if no packets were lost
   // and there are still crypto packets in flight.
-  else if (crypto packets are in flight):
+  else if (crypto packets are in flight &&
+           crypto data to send):
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: I still don't get this. If crypto packets are in flight, doesn't that mean there is unacked crypto data to send? Why is this second predicate useful?

Comment (ianswett) [reply draft-ietf-quic-recovery.md]: Crypto packets could be in flight because of a timer based retransmission, but the original was ACked.

Comment (ianswett) [reply draft-ietf-quic-recovery.md]: See issue #1964

Comment (janaiyengar) [reply draft-ietf-quic-recovery.md]: Ah, ok. How about just "unacked crypto data exists"? Is that enough to capture it?

Comment (janaiyengar) [reply draft-ietf-quic-recovery.md]: I meant using just this one condition instead of the two that you have.

Comment (Ralith) [reply draft-ietf-quic-recovery.md]: I agree that this wording is confusing; in particular, it's not obvious that "to send" means "not yet acknowledged," except perhaps by process of elimination.

Comment (ianswett) [reply draft-ietf-quic-recovery.md]: Sure, I took Jana's suggestion.
